mantis-runtime/src/main/java/io/mantisrx/runtime/executor/WorkerPublisherRemoteObservable.java

```diff
@@ -225,6 +104,52 @@ public void blockUntilServerShutdown() {
         }
     }
 
+    private <K> void startKeyValueStage(KeyValueStageConfig<?, K, T> stage, Observable<Observable<T>> toServe) {
+
+        if (runNewW2WserverGroups(jobName)) {
+            logger.info("Modern server setup for name: " + name + " type: Keyedstage");
+
+            long expiryTimeInSecs = Long.MAX_VALUE;
+            if (stage instanceof KeyToKey) {
+                expiryTimeInSecs = ((KeyToKey) stage).getKeyExpireTimeSeconds();
+            } else if (stage instanceof ScalarToKey) {
+                expiryTimeInSecs = ((ScalarToKey) stage).getKeyExpireTimeSeconds();
+            }
+
+            Func1<T, byte[]> valueEncoder = t1 -> stage.getOutputCodec().encode(t1);
+
+            Func1<K, byte[]> keyEncoder = t1 -> stage.getOutputKeyCodec().encode(t1);
+
+
+            ServerConfig<KeyValuePair<K, T>> config = new ServerConfig.Builder<KeyValuePair<K, T>>()
+                .name(name)
+                .port(serverPort)
+                .metricsRegistry(MetricsRegistry.getInstance())
+                .numQueueConsumers(numConsumerThreads())
+                .maxChunkSize(maxChunkSize())
+                .maxChunkTimeMSec(maxChunkTimeMSec())
+                .bufferCapacity(bufferCapacity())
+                .useSpscQueue(useSpsc())
+                .router(Routers.consistentHashingLegacyTcpProtocol(jobName, keyEncoder, valueEncoder))
+                .build();
+
+            final LegacyTcpPushServer<KeyValuePair<K, T>> modernServer;
+            if (stage instanceof ScalarToGroup || stage instanceof GroupToGroup) {
+                modernServer = PushServers.infiniteStreamLegacyTcpNestedMantisGroup(
+                    config, (Observable) toServe, expiryTimeInSecs, keyEncoder,
+                    io.reactivex.mantis.network.push.HashFunctions.ketama());
+            } else { // ScalarToKey or KeyTKey
+                modernServer = PushServers.infiniteStreamLegacyTcpNestedGroupedObservable(
+                    config, (Observable) toServe, expiryTimeInSecs, keyEncoder,
+                    io.reactivex.mantis.network.push.HashFunctions.ketama());
+            }
+
+            modernServer.start();
```

Comment:
nit: why not start modernServer as part of LegacyServer::start? That way the lifecycle of the legacyserver gets fully utilized.

Reply:
I didn't look too much into the internals here to make that change. I'm planning on leaving this be for now and we can solve it later.

Reply:
fair enough. Can you just add a TODO in the code?

Reply:
fixed it now. thanks!